internal/testing/e2e/expectCertifyLegalQ1.json

```diff
@@ -0,0 +1,29 @@
+{
```

Comment:
The output will be this based on the new query:

```
{
  "data": {
    "CertifyLegal": [
      {
        "id": "certify_legals:9fb7c1e4-90fe-5233-8600-3b52109b42a2",
        "subject": {
          "__typename": "Package",
          "id": "package_types:maven",
          "type": "maven",
          "namespaces": [
            {
              "id": "package_namespaces:mavenguac-split-@@org.apache.logging.log4j",
              "namespace": "org.apache.logging.log4j",
              "names": [
                {
                  "id": "package_names:daa4fc66-77d7-5a65-96db-df05b7e241df",
                  "name": "log4j-core",
                  "versions": [
                    {
                      "id": "package_versions:a4da45f4-8e28-55ef-aaa1-194a09100cc8",
                      "purl": "pkg:maven/org.apache.logging.log4j/log4j-core@2.8.1",
                      "version": "2.8.1",
                      "qualifiers": [],
                      "subpath": ""
                    }
                  ]
                }
              ]
            }
          ]
        },
        "declaredLicense": "NONE",
        "declaredLicenses": [
          {
            "id": "licenses:62602d05-7fca-58cb-95b2-023ca5f15d06",
            "name": "GPL-2.0-only",
            "inline": null,
            "listVersion": "3.18"
          },
          {
            "id": "licenses:84ba038d-4a9a-5e46-b823-7676377da1e4",
            "name": "Zlib",
            "inline": null,
            "listVersion": "3.18"
          },
          {
            "id": "licenses:da114a2d-ae19-5d2b-a83b-639cae65b861",
            "name": "BSD-3-Clause",
            "inline": null,
            "listVersion": "3.18"
          }
        ],
        "discoveredLicense": "NONE",
        "discoveredLicenses": [
          {
            "id": "licenses:62602d05-7fca-58cb-95b2-023ca5f15d06",
            "name": "GPL-2.0-only",
            "inline": null,
            "listVersion": "3.18"
          },
          {
            "id": "licenses:84ba038d-4a9a-5e46-b823-7676377da1e4",
            "name": "Zlib",
            "inline": null,
            "listVersion": "3.18"
          },
          {
            "id": "licenses:da114a2d-ae19-5d2b-a83b-639cae65b861",
            "name": "BSD-3-Clause",
            "inline": null,
            "listVersion": "3.18"
          }
        ],
        "attribution": "",
        "justification": "Found in SPDX document.",
        "timeScanned": "2022-10-19T17:29:41.243562Z",
        "origin": "file:///../guac-data/docs/spdx/spdx_vuln.json",
        "collector": "FileCollector"
      },
      {
        "id": "certify_legals:3698167c-6db1-563c-833a-430f023431b0",
        "subject": {
          "__typename": "Package",
          "id": "package_types:maven",
          "type": "maven",
          "namespaces": [
            {
              "id": "package_namespaces:mavenguac-split-@@org.apache.logging.log4j",
              "namespace": "org.apache.logging.log4j",
              "names": [
                {
                  "id": "package_names:daa4fc66-77d7-5a65-96db-df05b7e241df",
                  "name": "log4j-core",
                  "versions": [
                    {
                      "id": "package_versions:a4da45f4-8e28-55ef-aaa1-194a09100cc8",
                      "purl": "pkg:maven/org.apache.logging.log4j/log4j-core@2.8.1",
                      "version": "2.8.1",
                      "qualifiers": [],
                      "subpath": ""
                    }
                  ]
                }
              ]
            }
          ]
        },
        "declaredLicense": "Apache-2.0",
        "declaredLicenses": [],
        "discoveredLicense": "Apache-2.0",
        "discoveredLicenses": [
          {
            "id": "licenses:bfb0efe5-4ba1-54b7-a210-31f3eac12b45",
            "name": "NOASSERTION",
            "inline": null,
            "listVersion": "UNKNOWN"
          }
        ],
        "attribution": "",
        "justification": "Retrieved from ClearlyDefined",
        "timeScanned": "2024-10-09T10:52:44.791488Z",
        "origin": "clearlydefined",
        "collector": "clearlydefined"
      }
    ]
  }
}
```

Reply:
We will have to ignore the `IDs` and `timeScanned` like you are doing.

Reply:
Thank you @pxp928 I will strip those fields!

Reply:
Hmm, when I run the command in my local on `guac-data-main` I get the following garphql error:
```
    {
      "message": "the requested element is null which the schema does not allow",
      "path": [
        "CertifyLegal",
        21
      ]
    },
```
Also, when I run the query inside a github workflow using [act](https://github.com/nektos/act), it fails with an empty response for some reason. I've updated to the result from above, so hopefully it will work in the gh action. I think I am missing something cc @pxp928

Reply:
Hey @robert-cronin so this is an actual bug you found. This issue is here in the keyvalue backend - https://github.com/guacsec/guac/blob/main/pkg/assembler/backends/keyvalue/certifyLegal.go#L550

It needs to check if the output is not nil and not append it if it is nil:
```
legal, err := c.legalIfMatch(ctx, filter, link)
if err != nil {
     return nil, gqlerror.Errorf("%v :: %v", funcName, err)
}
if legal != nil {
    out = append(out, legal)
}
```

Once that is fixed, it will return the correct output for this query!

Reply:
This issue does not happen in our postgres backend so you should not have any errors there!

Reply:
oh wow, glad I could stumble on a bug to solve as well!
Thank you @pxp928 I have added the fix here: https://github.com/guacsec/guac/pull/2193

Reply:
It looks like there is a flaky PkgQ3 test here: https://github.com/guacsec/guac/actions/runs/11285128425/job/31387424106?pr=2168#step:9:18224
Not sure how to retrigger the test, is there a way to retrigger them? cc @pxp928

Reply:
re-triggered!

Reply:
Thanks @pxp928
I've noticed that sometimes the query returns two matching (apparently equal) subjects; one collected from the files and the other collected from clearly defined. Example with both (second object is the one collected from CD):

```
{
  "CertifyLegal": [
    {
      "subject": {
        "__typename": "Package",
        "type": "maven",
        "namespaces": [
          {
            "namespace": "org.apache.logging.log4j",
            "names": [
              {
                "name": "log4j-core",
                "versions": [
                  {
                    "purl": "pkg:maven/org.apache.logging.log4j/log4j-core@2.8.1",
                    "version": "2.8.1",
                    "qualifiers": [],
                    "subpath": ""
                  }
                ]
              }
            ]
          }
        ]
      },
      "declaredLicense": "NONE",
      "discoveredLicense": "NONE",
      "attribution": "",
      "justification": "Found in SPDX document.",
      "collector": "FileCollector"
    },
    {
      "subject": {
        "__typename": "Package",
        "type": "maven",
        "namespaces": [
          {
            "namespace": "org.apache.logging.log4j",
            "names": [
              {
                "name": "log4j-core",
                "versions": [
                  {
                    "purl": "pkg:maven/org.apache.logging.log4j/log4j-core@2.8.1",
                    "version": "2.8.1",
                    "qualifiers": [],
                    "subpath": ""
                  }
                ]
              }
            ]
          }
        ]
      },
      "declaredLicense": "Apache-2.0",
      "discoveredLicense": "Apache-2.0",
      "attribution": "",
      "justification": "Retrieved from ClearlyDefined",
      "collector": "clearlydefined"
    }
  ]
}
```

To resolve this and only target the CD collector, I have added a gql filter to select the one from CD. Seems to be working now in the pipeline.